rpc/src/client/transport/router.rs

```diff
@@ -17,36 +19,54 @@ pub struct SubscriptionRouter {
     // their result channels. Used for publishing events relating to a specific
     // query.
     subscriptions: HashMap<String, HashMap<String, SubscriptionTx>>,
+
+    // A map of subscription ids to their queries
+    sub_to_query: HashMap<String, String>,
```

Comment:
Is this an optimization to allow for reverse lookup of queries from their subscription IDs?

Reply:
Yes, maybe I should i prove the naming and the comment if that wasn't so clear.

Reply:
It's more the optimization I was wondering about. I generally prefer to have a "single source of truth" when it comes to the data being managed, because data duplication results in additional complexity. In this case, what's the computational cost of doing a query lookup from an ID from the `subscriptions` data structure?

We could do something like:

```rust
/// Returns the query string associated with the subscription with the given ID.
pub fn get_query_by_id(&self, id: &str) -> Option<String> {
    self.subscriptions.iter().find_map(|(query, subs)| {
        if subs.contains_key(id) {
            Some(query.clone())
        } else {
            None
        }
    })
}
```

Reply:
I share your concern about the single source of truth but I am not a big fan of iterating over the map to find the matching entry. How about we can extract this into a generic data structure with proper invariants enforced by the `contracts` create, to guarantee that the data does not go out of sync? Maybe we can even prove it correct with the Rust Stainless front end ;)

Reply:
Just so I understand, are there any drawbacks of iterating through the map apart from the additional computational overhead?

For me, the optimized version provides `O(1)` lookup time at the cost of additional data structure management complexity (the two data structures become implicitly intertwined without any compile-time or runtime guarantees as to their coherence, and potentially complicating future changes). It introduces this data structure complexity for an edge case that is, at least partially, a bug in Tendermint.

From what I can see, the unoptimized, but simpler version provides `O(N)` lookup time (since each ID lookup is `O(1)`), where `N` = no. of unique queries for which there are subscriptions. How many subscription queries do we anticipate an RPC client will be creating simultaneously? 10's, 100's, 1000's?

Reply:
Good point, we will probably only ever have a handful of queries per RPC client (currently 2), so yeah I agree with you that this optimization is perhaps unnecessary 👍